src/connectors/wire/WireServer.cpp

```diff
@@ -1,40 +1,95 @@
 #include <cucumber-cpp/internal/connectors/wire/WireServer.hpp>
+#include <boost/filesystem/operations.hpp>
 
 namespace cucumber {
 namespace internal {
 
+using namespace boost::asio::ip;
+#if defined(BOOST_ASIO_HAS_LOCAL_SOCKETS)
+using namespace boost::asio::local;
+#endif
+
 SocketServer::SocketServer(const ProtocolHandler *protocolHandler) :
     protocolHandler(protocolHandler),
-    ios(),
-    acceptor(ios) {
+    ios() {
 }
 
-void SocketServer::listen(const port_type port) {
-    tcp::endpoint endpoint(tcp::v4(), port);
+template <typename Protocol, typename Service>
+void SocketServer::DoListen(basic_socket_acceptor<Protocol, Service>& acceptor,
+        const typename Protocol::endpoint& endpoint) {
     acceptor.open(endpoint.protocol());
-    acceptor.set_option(tcp::acceptor::reuse_address(true));
-    acceptor.set_option(tcp::no_delay(true));
+    acceptor.set_option(typename Protocol::acceptor::reuse_address(true));
     acceptor.bind(endpoint);
     acceptor.listen(1);
 }
 
-SocketServer::port_type SocketServer::listenPort() const {
-  const tcp::endpoint ep(acceptor.local_endpoint());
-  return ep.port();
-}
-
-void SocketServer::acceptOnce() {
-    tcp::iostream stream;
+template <typename Protocol, typename Service>
+void SocketServer::DoAcceptOnce(basic_socket_acceptor<Protocol, Service>& acceptor) {
+    typename Protocol::iostream stream;
     acceptor.accept(*stream.rdbuf());
     processStream(stream);
 }
 
-void SocketServer::processStream(tcp::iostream &stream) {
+void SocketServer::processStream(std::iostream &stream) {
     std::string request;
     while (getline(stream, request)) {
         stream << protocolHandler->handle(request) << std::endl << std::flush;
     }
 }
 
+TCPSocketServer::TCPSocketServer(const ProtocolHandler *protocolHandler) :
+    SocketServer(protocolHandler),
+    acceptor(ios) {
+}
+
+void TCPSocketServer::listen(const port_type port) {
+    DoListen(acceptor, tcp::endpoint(tcp::v4(), port));
+    acceptor.set_option(tcp::no_delay(true));
+}
+
+tcp::endpoint TCPSocketServer::listenEndpoint() const {
+    return acceptor.local_endpoint();
+}
+
+void TCPSocketServer::acceptOnce() {
+    DoAcceptOnce(acceptor);
+}
+
+#if defined(BOOST_ASIO_HAS_LOCAL_SOCKETS)
+UnixSocketServer::UnixSocketServer(const ProtocolHandler *protocolHandler) :
+    SocketServer(protocolHandler),
+    acceptor(ios) {
+}
+
+void UnixSocketServer::listen(const std::string& unixPath) {
+    removeUnixSocket();
+
+    if (boost::filesystem::status(unixPath).type() == boost::filesystem::socket_file)
+        boost::filesystem::remove(unixPath);
+
+    DoListen(acceptor, stream_protocol::endpoint(unixPath));
+    socketToRemove = unixPath;
```

Comment:
The removal of the socket file on startup is actually pretty common practice for unix sockets (I'm not sure whether it's more or less prevalent than failing on startup though). Necessarily so, because something like `reuse_addr` (permit binds on existing endpoints that nothing is currently bound to) doesn't exist for unix sockets. Not doing this requires manual intervention, which for server startup is probably not desired. I do have the file-type checking there to prevent accidental deletion of files containing data though.

Note that interference with other programs shouldn't happen if the socket-file-name is chosen wisely (which is a task for the user in this case, choosing a unique name) and interfering with other instances of the same cucumber-cpp program can only happen if it's started twice, which is also something a user shouldn't do.

I missed the removal of `socketToRemove`, added it now.